### PR TITLE
Add App Native Auth support

### LINKED ISSUE #21916: Add app native authentication support for UAEPassAuthenticator
**Describe the improvement:**

Currently only few authenticators support app native authentication and UAE Pass [1] is not one of them.

**Additional context:**
- Product Version: IS 7.0.0

---
[1] https://github.com/wso2-extensions/identity-outbound-auth-uaepass/blob/main/components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java#L81

### PR DESCRIPTION
**Pull request by janakamarasena**
## Purpose
Resolves https://github.com/wso2/product-is/issues/21916

### CODE REVIEW

File: components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticatorConstants.java [hunk 1]
```diff
@@ -49,7 +65,14 @@ public enum ErrorMessages {
         UAEPASS_AUTHN_FAILED_ACCESS_TOKEN_BUILD_FAILURE("UAEPass-65011", "UAEPass Authentication" +
                 "Exception while building access token request with the request body"),
         UAEPASS_AUTHN_FAILED_ABSOLUTE_URL_BUILD_FAILURE("UAEPass-65012", "Error occurred while " +
-                "extracting the absolute public URL from the browser");
+                "extracting the absolute public URL from the browser"),
```
Comment (pamodaaw) [inline components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticatorConstants.java:68]:
```suggestion
                "extracting the absolute public URL from the browser."),
```

File: components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticatorConstants.java [hunk 2]
```diff
@@ -49,7 +65,14 @@ public enum ErrorMessages {
...
+        ACCESS_TOKEN_EMPTY("UAEPass-65013", "Access token is empty"),
```
Comment (pamodaaw) [inline components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticatorConstants.java]:
```suggestion
        ACCESS_TOKEN_EMPTY("UAEPass-65013", "Access token is empty."),
```

File: components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticatorConstants.java [hunk 3]
```diff
@@ -49,7 +65,14 @@ public enum ErrorMessages {
...
+        RETRIEVING_AUTHENTICATOR_PROPERTIES_FAILED("UAEPass-65014", "Error while retrieving properties. "
+                + "Authenticator properties cannot be null."),
+        USER_ID_NOT_FOUND_IN_ID_TOKEN("UAEPass-65015", "Cannot find the userId from the id_token sent "
+                + "by the federated IDP."),
+        NO_REGISTERED_IDP_FOR_ISSUER("UAEPass-65016", "No registered IdP found for the issuer: %s"),
```
Comment (pamodaaw) [inline components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticatorConstants.java:74]:
```suggestion
        NO_REGISTERED_IDP_FOR_ISSUER("UAEPass-65016", "No registered IdP found for the issuer: %s."),
```

File: components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticatorConstants.java [hunk 4]
```diff
@@ -49,7 +65,14 @@ public enum ErrorMessages {
...
+        JWT_TOKEN_VALIDATION_FAILED("UAEPass-65017", "Error while validating the JWT token");
```
Comment (pamodaaw) [inline components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticatorConstants.java]:
```suggestion
        JWT_TOKEN_VALIDATION_FAILED("UAEPass-65017", "Error while validating the JWT token.");
```

File: components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java [hunk 5]
```diff
@@ -92,7 +119,14 @@ public class UAEPassAuthenticator extends AbstractApplicationAuthenticator
...
 
-        return UAEPassAuthenticatorConstants.UAE.LOGIN_TYPE.equals(getLoginType(request));
+        boolean canHandle;
+        if (isNativeSDKBasedFederationCall(request)) {
+            canHandle = true;
+        } else {
+            canHandle = UAEPassAuthenticatorConstants.UAE.LOGIN_TYPE.equals(getLoginType(request));
+        }
+
+        return canHandle;
```
Comment (pamodaaw) [inline components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java]: Can't we just return the value without introducing the new variable `canHandle`

File: components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java [hunk 6]
```diff
@@ -340,24 +398,32 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+                throw new AuthenticationFailedException(UAEPassAuthenticatorConstants.ErrorMessages.
+                        USER_ID_NOT_FOUND_IN_ID_TOKEN.getMessage());
             }
             authenticatedUser = AuthenticatedUser.createFederateAuthenticatedUserFromSubjectIdentifier
                     (authenticatedUserId);
             authenticatedUser.setUserAttributes(claims);
             context.setSubject(authenticatedUser);
 
         } catch (UAEPassAuthnFailedException e) {
+            setAuthenticatorMessageToContext(UAEPassAuthenticatorConstants.ErrorMessages.
```
Comment (pamodaaw) [inline components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java:412]: Why are we setting the error message to the context?

File: components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java [hunk 7]
```diff
@@ -972,4 +1048,244 @@ private String getFileConfigValue(String fileConfigKey) {
...
+    private String getStateParameter(HttpServletRequest request, AuthenticationContext context) {
+
+        String state;
+        if (isAPIBasedAuthenticationFlow(request, context)) {
+            state = UUID.randomUUID() + "," + UAEPassAuthenticatorConstants.UAE.LOGIN_TYPE;
+        } else {
+            state = context.getContextIdentifier() + "," + UAEPassAuthenticatorConstants.UAE.LOGIN_TYPE;
+        }
+
+        return state;
```
Comment (pamodaaw) [inline components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java]: can't we just return the value within the if condition

File: components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java [hunk 8]
```diff
@@ -972,4 +1048,244 @@ private String getFileConfigValue(String fileConfigKey) {
...
+     * @param context Authentication context.
+     * @param scope Space separated scopes.
+     */
+    private void addScopeToContext(AuthenticationContext context, String scope) {
+
+        if (StringUtils.isNotBlank(scope)) {
+            context.setProperty(getName() + UAEPassAuthenticatorConstants.SCOPE_PARAM_SUFFIX, scope);
+        }
+
+    }
```
Comment (pamodaaw) [inline components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java]:
```suggestion
        if (StringUtils.isNotBlank(scope)) {
            context.setProperty(getName() + UAEPassAuthenticatorConstants.SCOPE_PARAM_SUFFIX, scope);
        }
    }
```

File: components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java (hunk 6 above)
Comment (janakamarasena) [reply components/uaepass-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/uaepass/UAEPassAuthenticator.java:412]: This is read from the app native auth code to send the error details in the api response.

**Review by pamodaaw — APPROVED** [review]

**Pull request merged**